src/neuroconv/basedatainterface.py

```diff
@@ -182,32 +185,51 @@ def run_conversion(
             Otherwise, all datasets will use default configuration settings.
         """
 
-        backend = _resolve_backend(backend, backend_configuration)
-        no_nwbfile_provided = nwbfile is None  # Otherwise, variable reference may mutate later on inside the context
+        appending_to_in_memory_nwbfile = nwbfile is not None
+        file_initially_exists = Path(nwbfile_path).exists() if nwbfile_path is not None else False
+        appending_to_disk_nwbfile = file_initially_exists and not overwrite
 
         if metadata is None:
             metadata = self.get_metadata()
+        self.validate_metadata(metadata=metadata, append_mode=appending_to_disk_nwbfile)
+
+        if not appending_to_disk_nwbfile:
+            if appending_to_in_memory_nwbfile:
+                self.add_to_nwbfile(nwbfile=nwbfile, metadata=metadata, **conversion_options)
+            else:
+                nwbfile = self.create_nwbfile(metadata=metadata, **conversion_options)
+
+            configure_and_write_nwbfile(
+                nwbfile=nwbfile,
+                output_filepath=nwbfile_path,
+                backend=backend,
+                backend_configuration=backend_configuration,
+            )
+
+        else:  # We are only using the context in append mode, see issue #1143
+
+            if appending_to_disk_nwbfile:
+                raise ValueError(
+                    "Cannot append to an existing file while also providing an in-memory NWBFile. "
+                    "Either set overwrite=True to replace the existing file, or remove the nwbfile parameter to append to the existing file on disk."
+                )
```

Comment:
I would prefer to raise this value error right at the beginning for clarity. Also, I think it should be 
```python
if appending_to_disk_nwbfile and appending_to_in_memory_nwbfile:
    raise ValueError(...)
```
right?

Reply:
The condition was wrong, I fixed it now. Thanks. 

I would like the error to be nested in the apend mode branch. I am also coming at this from the clarity angle but with a different reasoning.

People who read the code should not be bothered with the complexities of the apend case unless they get to that branch. In most cases, `run_conversion` is working as a `write_to_nwbfile` (which I think should be a method!) and I don't want the first reading of the function to introduce unrelated complexities.

Reply:
Those are good points. I will move it. 


I still have the lingering feeling of wanting to push the complexities of appending mode outside of the happy path but maybe I should push for a `write_to_nwbfile` functionality outside of this PR. I think should be easier to read, use, and document for both developers and users.
  
`run_conversion`, like the context manager here, has [too many responsibilities](https://dev.to/eidorianavi/functions-and-the-single-responsibility-principle-48ae) which leads [to too many arguments ](https://dev.to/thinkster/code-smell-too-many-parameters-435e) an anti-pattern and a code smell.

Reply:
> I still have the lingering feeling of wanting to push the complexities of appending mode outside of the happy path but maybe I should push for a `write_to_nwbfile` functionality outside of this PR. I think should be easier to read, use, and document for both developers and users.
> 
> `run_conversion`, like the context manager here, has [too many responsibilities](https://dev.to/eidorianavi/functions-and-the-single-responsibility-principle-48ae) which leads [to too many arguments ](https://dev.to/thinkster/code-smell-too-many-parameters-435e) an anti-pattern and a code smell.

Yep, isolating the complexity of append mode makes sense to me. I look forward to seeing those changes.